### PR TITLE
Streaming processing

### PR DESCRIPTION
**Pull request by jeremydeanlakey**
This is mostly about reading and writing in a streaming way, although there is some other improvements thrown in. Changes also reflects the other pending pull request.

### CODE REVIEW

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java [hunk 1]
```diff
@@ -40,41 +41,104 @@ public boolean contains(BlueKey key) throws BlueDbException {
...
+			@Override
+			public void process(BlueObjectInput<BlueEntity<T>> input, BlueObjectOutput<BlueEntity<T>> output) throws BlueDbException {
+				BlueEntity<T> newEntity = new BlueEntity<T>(newKey, newValue);
+				while (input.hasNext()) {
+					BlueEntity<T> iterEntity = input.next();
+					BlueKey iterKey = iterEntity.getKey();
+					if (iterKey.equals(newKey)) {
+						output.write(newEntity);
+						newEntity = null;
+					} else if (newEntity != null && iterKey.getGroupingNumber() > groupingNumber) {
```
Comment (d-w-johnson) [inline BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: So if I delete something then try to update it then the object will be resurrected? Or I guess in order to have the newValue that means that we already found the value once and applied the update lambda to it right? Why do we end up needing this branch?

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java [hunk 2]
```diff
@@ -40,41 +41,104 @@ public boolean contains(BlueKey key) throws BlueDbException {
...
-		try (BlueReadLock<Path> readLock = getReadLockFor(key)) {
-			if (readLock != null) {
-				entities = fetch(readLock);
-				file = readLock.getKey().toFile();
-			} else {
-				entities = new ArrayList<>();
-				file = 	getPathFor(key, 1).toFile();
+	public void update(BlueKey newKey, T newValue) throws BlueDbException {
+		long groupingNumber = newKey.getGroupingNumber();
+		modifyChunk(groupingNumber, new Processor<T>() {
```
Comment (d-w-johnson) [inline BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: I don't like creating a new anonymous inner class every time for all of these processes. I'd extract the contents into a class or method and use a lambda or method reference. Or at lease create a private `updateProcessor` member that is an anonymous inner class and can be used every time.

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java [hunk 3]
```diff
@@ -41,6 +43,10 @@ public boolean contains(BlueKey key) throws BlueDbException {
 		return get(key) != null;
 	}
 
+	interface Processor<X extends Serializable> {
```
Comment (d-w-johnson) [inline BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: I also think this deserves its own class

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java [hunk 4]
```diff
@@ -28,10 +28,19 @@
...
+	private final LockManager<Path> lockManager;
 
 	public Segment(Path segmentPath, FileManager fileManager) {
 		this.segmentPath = segmentPath;
 		this.fileManager = fileManager;
+		lockManager = fileManager.getLockManager();
+	}
+
+	// for testing only
+	protected Segment() {
```
Comment (d-w-johnson) [inline BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: How do you feel about making a static method called `createTestSegment()` or something like that? Then you'd easily be able to make a test instance of the class without the risk of someone trying to use it in non-test environments.

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java (hunk 1 above)
Comment (jeremydeanlakey) [reply BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: removed recreate on update

**Pull request merged**